### PR TITLE
ZenBull shutdown withdraw

### PR DESCRIPTION
**Pull request by aleone**
# Task: ZenBull shutdown withdraw

## Description

Withdraws remaining ZenBull funds from Crab prior to shutdown, which can be claimed by users by providing their ZenBull tokens

Fixes (linear-task)

## Type of change

- [x] New feature
- [ ] Bug fix
- [ ] Testing code
- [ ] Document update or config files

### CODE REVIEW

**Review by alpinechicken — APPROVED** [review]
lgtm

File: packages/zen-bull-vault/src/ShutdownEmergencyWithdraw.sol [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+     */
+    receive() external payable {
+        require(msg.sender == weth, "Can't receive ETH from sender");
+    }
+
+    /**
+     * @notice withdraw from all remaining Crab in ZenBull strategy
+     * @dev only owner can call this function
+     */
+    function shutdownEmergencyWithdraw() external onlyOwner {
```
Comment (nikkaroraa) [inline packages/zen-bull-vault/src/ShutdownEmergencyWithdraw.sol:107]: better natspec can be helpful i feel:

```solidity
/**
 * @notice Emergency function to withdraw remaining Crab tokens from ZenBull and unwind positions
 * @dev Only callable by owner. Must be called before users can claim their share via claimZenBullRedemption
 * @dev ZenBull holders will be able to redeem their proportional share of the withdrawn WETH
 * @dev This function:
 * 1. Calculates how many wPowerPerp tokens are needed to redeem the Crab position
 * 2. Takes wPowerPerp tokens from msg.sender (the person helping unwind)
 * 3. Redeems Crab tokens for WETH
 * 4. Compensates msg.sender with WETH equal to their wPowerPerp value
 * 5. Keeps remaining WETH for ZenBull holders to claim
 */
 ```

File: packages/zen-bull-vault/src/ShutdownEmergencyWithdraw.sol [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        require(zenBullTotalSupplyAtRedemption == 0, "ZenBull total supply set");
+        uint256 crabToRedeem = IZenBullStrategy(zenBull).getCrabBalance();
+        (, uint256 wPowerPerpInCrab) = IZenBullStrategy(zenBull).getCrabVaultDetails();
+
+        uint256 ethIndexPrice =
+            IOracle(oracle).getTwap(ethUsdcPool, weth, usdc, TWAP_PERIOD, true).div(INDEX_SCALE);
+
+        uint256 wPowerPerpToRedeem =
+            crabToRedeem.wmul(wPowerPerpInCrab).wdiv(IERC20(crab).totalSupply());
+        uint256 ethValueInWPowerPerp = wPowerPerpToRedeem.wmul(ethIndexPrice).wmul(
```
Comment (nikkaroraa) [inline packages/zen-bull-vault/src/ShutdownEmergencyWithdraw.sol:117]: isn't this `wPowerPerpValueInEth`?

File: packages/zen-bull-vault/src/ShutdownEmergencyWithdraw.sol [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+            msg.sender,
+            crabToRedeem,
+            wPowerPerpToRedeem,
+            totalPayout,
+            ethValueInWPowerPerp,
+            wethAtRedemption
+        );
+    }
+
+    function claimZenBullRedemption(uint256 _amountToRedeem) external {
```
Comment (nikkaroraa) [inline packages/zen-bull-vault/src/ShutdownEmergencyWithdraw.sol:143]: natspec:

```solidity
/**
 * @notice Allows ZenBull token holders to claim their proportional share of WETH after emergency shutdown
 * @dev After shutdownEmergencyWithdraw has been called, ZenBull holders can redeem their tokens for WETH
 * @dev The WETH amount received is proportional to their share of total ZenBull supply at time of shutdown
 * @param _amountToRedeem Amount of ZenBull tokens to redeem for WETH
 * @dev Formula: wethOwed = (amountToRedeem / zenBullTotalSupplyAtRedemption) * wethAtRedemption
 */
 ```

File: packages/zen-bull-vault/test/e2e/ShutdownEmergencyWithdrawTest.t.sol [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+    }
+
+    function testShutdownEmergencyWithdraw() public {
+        // get initial crab balance in zen bull strategy
+        uint256 initialZenBullCrabBalance = ZenBullStrategy(ZEN_BULL).getCrabBalance();
+
+        (uint256 ethInCrab, uint256 wPowerPerpInCrab) =
+            ZenBullStrategy(ZEN_BULL).getCrabVaultDetails();
+
+        uint256 ethShare = initialZenBullCrabBalance.wdiv(IERC20(CRAB).totalSupply());
```
Comment (nikkaroraa) [inline packages/zen-bull-vault/test/e2e/ShutdownEmergencyWithdrawTest.t.sol:82]: doesn't change anything - but this is more like crabShare

File: packages/zen-bull-vault/test/e2e/ShutdownEmergencyWithdrawTest.t.sol [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+    }
+
+    function testClaimZenBullRedemption() public {
+        // get initial crab balance in zen bull strategy
+        uint256 initialZenBullCrabBalance = ZenBullStrategy(ZEN_BULL).getCrabBalance();
+
+        (uint256 ethInCrab, uint256 wPowerPerpInCrab) =
+            ZenBullStrategy(ZEN_BULL).getCrabVaultDetails();
+
+        uint256 ethShare = initialZenBullCrabBalance.wdiv(IERC20(CRAB).totalSupply());
```
Comment (nikkaroraa) [inline packages/zen-bull-vault/test/e2e/ShutdownEmergencyWithdrawTest.t.sol:149]: same comment, more like crabShare but doesn't affect anything

**Review by nikkaroraa — APPROVED** [review]
overall functionality looks good. 

added some comments on variable names and natspec description.

File: packages/zen-bull-vault/test/e2e/ShutdownEmergencyWithdrawTest.t.sol (hunk 4 above)
Comment (aleone) [reply packages/zen-bull-vault/test/e2e/ShutdownEmergencyWithdrawTest.t.sol:82]: changed to share

File: packages/zen-bull-vault/test/e2e/ShutdownEmergencyWithdrawTest.t.sol (hunk 5 above)
Comment (aleone) [reply packages/zen-bull-vault/test/e2e/ShutdownEmergencyWithdrawTest.t.sol:149]: changed to share

File: packages/zen-bull-vault/src/ShutdownEmergencyWithdraw.sol (hunk 3 above)
Comment (aleone) [reply packages/zen-bull-vault/src/ShutdownEmergencyWithdraw.sol:143]: updated

File: packages/zen-bull-vault/src/ShutdownEmergencyWithdraw.sol (hunk 1 above)
Comment (aleone) [reply packages/zen-bull-vault/src/ShutdownEmergencyWithdraw.sol:107]: updated

File: packages/zen-bull-vault/src/ShutdownEmergencyWithdraw.sol (hunk 2 above)
Comment (aleone) [reply packages/zen-bull-vault/src/ShutdownEmergencyWithdraw.sol:117]: yeah it is, changed

File: packages/zen-bull-vault/src/ShutdownEmergencyWithdraw.sol [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+    /**
+     * @dev receive ETH
+     */
+    receive() external payable {
+        require(msg.sender == weth, "Can't receive ETH from sender");
+    }
+
+    /**
+     * @notice withdraw from all remaining Crab in ZenBull strategy
```
[the remark this answers is not in the document]
Comment (aleone) [reply packages/zen-bull-vault/src/ShutdownEmergencyWithdraw.sol:104]: not change as used updated natspec

**Pull request merged**